src/main/java/com/slack/kaldb/server/Kaldb.java

```diff
@@ -50,104 +47,109 @@ public Kaldb(Path configFilePath) throws IOException {
       KaldbLocalQueryService<LogMessage> searcher =
           new KaldbLocalQueryService<>(indexer.getChunkManager());
-      GrpcServiceBuilder searchBuilder =
-          GrpcService.builder().addService(searcher).enableUnframedRequests(true);
-      sb.service(searchBuilder.build());
-      sb.annotatedService(new ElasticsearchApiService(searcher));
-
       final int serverPort = KaldbConfig.get().getIndexerConfig().getServerPort();
-      addManagementEndpoints(sb, serverPort);
-
-      Server server = sb.build();
-
-      CompletableFuture<Void> serverFuture = server.start();
-      serverFuture.join();
-      LOG.info("Started indexer server on port: {}", serverPort);
-
-      // TODO: Instead of passing in the indexer, consider creating an interface or make indexer of
-      // subclass of this class?
-      indexer.start();
+      ArmeriaService armeriaService =
+          new ArmeriaService(serverPort, prometheusMeterRegistry, searcher, "armeriaIndexService");
+      services.add(armeriaService);
     }
 
     if (roles.contains(KaldbConfigs.NodeRole.QUERY)) {
-      ServerBuilder sb = Server.builder();
-
-      GrpcServiceBuilder searchBuilder =
-          GrpcService.builder()
-              .addService(new KaldbDistributedQueryService())
-              .enableUnframedRequests(true);
-      sb.service(searchBuilder.build());
-
+      KaldbDistributedQueryService searcher = new KaldbDistributedQueryService();
       final int serverPort = KaldbConfig.get().getQueryConfig().getServerPort();
-      addManagementEndpoints(sb, serverPort);
-
-      Server server = sb.build();
-
-      CompletableFuture<Void> serverFuture = server.start();
-      serverFuture.join();
-      LOG.info("Started query server on port: {}", serverPort);
+      ArmeriaService armeriaService =
+          new ArmeriaService(serverPort, prometheusMeterRegistry, searcher, "armeriaQueryService");
+      services.add(armeriaService);
     }
 
-    // TODO: On CTRL-C shut down the process cleanly. Ensure no write locks in indexer. Guava
-    // ServiceManager?
+    return services;
   }
 
-  private LoggingServiceBuilder getLoggingServiceBuilder() {
-    return LoggingService.builder()
-        // Not logging any successful response, say prom scraping /metrics every 30 seconds at INFO
-        .successfulResponseLogLevel(LogLevel.DEBUG)
-        .failureResponseLogLevel(LogLevel.ERROR)
-        // Remove all headers to be sure we aren't leaking any auth/cookie info
-        .requestHeadersSanitizer((ctx, headers) -> DefaultHttpHeaders.EMPTY_HEADERS);
+  public static ServiceManager.Listener getServiceManagerListener() {
+    return new ServiceManager.Listener() {
+      @Override
+      public void failure(Service service) {
+        LOG.error(
+            String.format(
+                "Service %s failed with cause %s",
+                service.getClass().toString(), service.failureCause().toString()));
+        LOG.error("Shutting down Kaldb due to failed service");
```

Comment:
I think it may be better to use the RuntimeHalter class here for shutdown. That way we would have a uniform way to handle fatals.

Reply:
Done!